src/execution/mod.rs

```diff
@@ -61,68 +56,20 @@ pub struct ExecutionContext {
 
 impl ExecutionContext {
     /// Construct a new `ExecutionContext` with the specified configuration
-    pub fn new(config: ExecutionConfig) -> Self {
-        let _ = &config; // avoid unused variable warning (it is used when some features are enabled)
-
-        let cfg = SessionConfig::default()
-            .with_batch_size(1)
-            .with_information_schema(true);
-
-        let runtime_env = RuntimeEnv::default();
-
-        #[cfg(feature = "s3")]
-        {
-            if let Some(object_store_config) = &config.object_store {
-                if let Some(s3_configs) = &object_store_config.s3 {
-                    info!("S3 configs exists");
-                    for s3_config in s3_configs {
-                        match s3_config.to_object_store() {
-                            Ok(object_store) => {
-                                info!("Created object store");
-                                if let Some(object_store_url) = s3_config.object_store_url() {
-                                    info!("Endpoint exists");
-                                    if let Ok(parsed_endpoint) = Url::parse(object_store_url) {
-                                        info!("Parsed endpoint");
-                                        runtime_env.register_object_store(
-                                            &parsed_endpoint,
-                                            Arc::new(object_store),
-                                        );
-                                        info!("Registered s3 object store");
-                                    }
-                                }
-                            }
-                            Err(e) => {
-                                log::error!("Error creating object store: {:?}", e);
-                            }
-                        }
-                    }
-                }
-            }
-        }
-
-        #[allow(unused_mut)] // used when deltalake is enabled
-        let mut state = SessionStateBuilder::new()
-            .with_default_features()
-            .with_runtime_env(runtime_env.into())
-            .with_config(cfg)
-            .build();
-
-        #[cfg(feature = "deltalake")]
-        {
-            state
-                .table_factories_mut()
-                .insert("DELTATABLE".to_string(), Arc::new(DeltaTableFactory {}));
+    pub fn try_new(config: &ExecutionConfig) -> Result<Self> {
+        let mut builder = DftSessionStateBuilder::new();
+        for extension in all_extensions() {
```

Comment:
I am quite pleased how this looks now. Adding new extensions should be pretty straightforward